plutus-errors/src/TH/Bootstrap.hs

```diff
@@ -0,0 +1,55 @@
+module TH.Bootstrap (
+    bootstrap
+    ) where
+
+import Language.Haskell.TH
+import Language.Plutus.Common
+import Data.Traversable
+import Data.List
+import Data.Function
+import Data.Foldable
+import qualified Data.Map as M
+
+-- | Given a list of errors (data constructors), it returns
+-- ErrorCode instances to bootstrap and paste the error codes
+bootstrap :: [Name] -> Q [Dec]
+bootstrap es = do
+    -- give them a unique number
+    let indexedEs = zip es [1..]
+    -- group them by their parent type
+    grouppedByParentType <- groupDataConstrs indexedEs
+    M.elems <$> M.traverseWithKey makeInstance grouppedByParentType
+
+    where
+      groupDataConstrs :: [IxError] -> Q (M.Map ParentName [IxError])
+      groupDataConstrs ns = foldlM (\ acc e -> do
+        DataConI _ _ parentName <- reify $ fst e
+        pure $ M.insertWith (++) parentName [e] acc
```

Comment:
For stuff like this I'd probably do a `for` to do the monadic bit and get all the `parentName`s, and then a normal fold over that, but not a big deal.

Reply:
As ever: I really think it's worth trying hard to avoid multi-line parenthesized expressions in the middle of a larger expression.